src/storage/v2/disk/label_index.cpp

```diff
@@ -0,0 +1,209 @@
+      kvstore_->db_->BeginTransaction(rocksdb::WriteOptions(), rocksdb::TransactionOptions()));
+}
+
+std::unique_ptr<rocksdb::Transaction> DiskLabelIndex::CreateAllReadingRocksDBTransaction() const {
+  auto tx = CreateRocksDBTransaction();
+  tx->SetReadTimestampForValidation(std::numeric_limits<uint64_t>::max());
+  return tx;
+}
+
+bool DiskLabelIndex::SyncVertexToLabelIndexStorage(const Vertex &vertex, uint64_t commit_timestamp) const {
+  auto disk_transaction = CreateRocksDBTransaction();
+
+  if (auto maybe_old_disk_key = utils::GetOldDiskKeyOrNull(vertex.delta); maybe_old_disk_key.has_value()) {
+    if (!disk_transaction->Delete(maybe_old_disk_key.value()).ok()) {
+      return false;
+    }
+  }
+
+  for (const LabelId index_label : index_) {
+    if (utils::Contains(vertex.labels, index_label)) {
+      if (!disk_transaction
+               ->Put(utils::SerializeVertexAsKeyForLabelIndex(index_label, vertex.gid),
+                     utils::SerializeVertexAsValueForLabelIndex(index_label, vertex.labels, vertex.properties))
+               .ok()) {
+        return false;
+      }
+    }
+  }
+
+  return CommitWithTimestamp(disk_transaction.get(), commit_timestamp);
+}
+
+/// TODO: this can probably be optimized
+bool DiskLabelIndex::ClearDeletedVertex(std::string_view gid, uint64_t transaction_commit_timestamp) const {
+  auto disk_transaction = CreateAllReadingRocksDBTransaction();
+
+  rocksdb::ReadOptions ro;
+  std::string strTs = utils::StringTimestamp(std::numeric_limits<uint64_t>::max());
+  rocksdb::Slice ts(strTs);
+  ro.timestamp = &ts;
+  auto it = std::unique_ptr<rocksdb::Iterator>(disk_transaction->GetIterator(ro));
+  for (it->SeekToFirst(); it->Valid(); it->Next()) {
+    if (std::string key = it->key().ToString(); gid == utils::ExtractGidFromLabelIndexStorage(key)) {
+      if (!disk_transaction->Delete(key).ok()) {
+        return false;
+      }
+    }
+  }
+
+  return CommitWithTimestamp(disk_transaction.get(), transaction_commit_timestamp);
+}
+
+bool DiskLabelIndex::DeleteVerticesWithRemovedIndexingLabel(uint64_t transaction_start_timestamp,
+                                                            uint64_t transaction_commit_timestamp) {
+  auto disk_transaction = CreateAllReadingRocksDBTransaction();
+
+  rocksdb::ReadOptions ro;
+  std::string strTs = utils::StringTimestamp(std::numeric_limits<uint64_t>::max());
+  rocksdb::Slice ts(strTs);
+  ro.timestamp = &ts;
+  bool deletion_success = true;
+  entries_for_deletion.WithLock([&deletion_success, transaction_start_timestamp,
+                                 disk_transaction_ptr = disk_transaction.get()](auto &tx_to_entries_for_deletion) {
+    if (auto tx_it = tx_to_entries_for_deletion.find(transaction_start_timestamp);
+        tx_it != tx_to_entries_for_deletion.end()) {
+      deletion_success = ClearTransactionEntriesWithRemovedIndexingLabel(*disk_transaction_ptr, tx_it->second);
+      tx_to_entries_for_deletion.erase(tx_it);
+    }
+  });
```

Comment:
I think you don't need to capture deleteion_success in lambda

```
bool deletion_success = entries_for_deletion.WithLock([transaction_start_timestamp,
                                 disk_transaction_ptr = disk_transaction.get()](auto &tx_to_entries_for_deletion) {
    if (auto tx_it = tx_to_entries_for_deletion.find(transaction_start_timestamp);
        tx_it != tx_to_entries_for_deletion.end()) {
        tx_to_entries_for_deletion.erase(tx_it);
      return ClearTransactionEntriesWithRemovedIndexingLabel(*disk_transaction_ptr, tx_it->second);
    }
    return false;
  });
``` 

you can do something like this so that lambda returns